miden-crypto/src/merkle/smt/historical/mod.rs

```diff
@@ -0,0 +1,500 @@
+//! Yields an `Smt` that can be accessed concurrently with historical views.
+//!
+//! ## Constraints
+//!
+//! * only one `Smt` shall ever be alive, multiple might blow up storage/memory
+//! * concurrent reads at a historic level must be acceptable
+//! * updates are relatively rare, every second~ish
+//! * the historical views have a very short livespan
+//!
+//! ## Nomenclature in use
+//!
+//! When the `latest` state is updated via `apply_mutations` we call that _forward mutation set_.
+//! During this process the _reversion mutation set_ is created. For efficiency we construct
+//! _reversion_-type (`Reversion`) from that.
+//!
+//! ## Implementation
+//!
+//! On every call to `apply_mutations` we create new `Reversion` and track that internally.
+//! We start with a bootstrap state `Smt`. For sake of brevity, `rev` is short for a `Reversion`
+//! instance.
+//!
+//! I.e. for the follwing call sequence
+//! ```text
+//! # [rev[n] .. rev[1] rev[0]] Smt
+//!
+//! #construction
+//! [] Smt # <- smt
+//!
+//! # apply_mtuations(fwd0)
+//! [rev(fwd0)] (smt + fwd0) = [rev(fwd0)] (smt')
+//!
+//! # apply_mtuations(fwd1)
+//! [rev(fwd0) rev(fwd1)] (smt' + fwd1) = [rev(fwd0) rev(fwd1)] (smt'')
+//! #                                      [2]       [1]         [0] <- historical index
+//! ```
+//!
+//! Now in practice we are very much interested in the key/value leaf pairs, so we cache them
+//! alongside with each reversion.
+
+use std::{
+    collections::{HashMap, VecDeque},
+    sync::{Arc, RwLock, RwLockReadGuard},
+    vec::Vec,
+};
+
+use crate::{
+    EMPTY_WORD, Word,
+    merkle::{
+        EmptySubtreeRoots, InnerNode, LeafIndex, MerkleError, MutationSet, NodeIndex, NodeMutation,
+        SMT_DEPTH, Smt, SmtLeaf, SmtProof, SparseMerklePath, smt::SparseMerkleTree,
+    },
+};
+
+#[cfg(test)]
+mod tests;
+
+#[allow(missing_docs)]
+#[derive(thiserror::Error, Debug)]
+pub enum HistoricalError {
+    #[error(transparent)]
+    MerkleError(#[from] MerkleError),
+}
+
+/// The offset enum representing the offset relative to the `latest` `Smt`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum HistoricalOffset {
+    ReversionsIdx(usize),
+    Latest,
+    TooAncient,
+}
+#[derive(Debug, Clone)]
+struct HistoricalReversion {
+    /// The root at this historical point
+    root: Word,
+    /// The mutations that were applied to reach this state (Arc'd for sharing)
+    /// This is a vector to support potential batching of mutations at a single historical point
+    mutations: Arc<MutationSet<SMT_DEPTH, Word, Word>>,
+    /// Calculating leaves is expense, so we keep a cache, based on the leaf digest.
```

Comment:
Observation: HistoricalReversion stores precomputed_leaves. Could we also store shared subtrees to reduce memory overhead? Our prototypes showed 10% memory savings potential in historical scenarios with many unchanged regions.